### PR TITLE
[TF2.0] Adding automatic mixed precision training (via graph rewrite) support to Resnet & Transformer

### PR DESCRIPTION
**Pull request by vinhngx**
Automatic Mixed Precision for TensorFlow has recently been introduced:
https://medium.com/tensorflow/automatic-mixed-precision-in-tensorflow-for-faster-ai-training-on-nvidia-gpus-6033234b2540

This PR adds automatic mixed precision training to Resnet and Transformer model via setting the flag combo `--dtype=fp16 --fp16_implementation=graph_rewrite`.

For example:

```
python3 keras_imagenet_main.py --dtype=fp16 --fp16_implementation=graph_rewrite
python3 transformer_main.py --dtype=fp16 --fp16_implementation=graph_rewrite
```

### CODE REVIEW

**Comment by vinhngx** [comment]
End-to-end training result on 8x NVIDIA V100 with `--automatic_mixed_precision` option:

```
Epoch 90/90
I0809 12:54:35.725960 139968602302272 keras_utils.py:85] BenchmarkMetric: {'global step':55700, 'time_taken': 46.783057,'examples_per_second': 4377.653220}
I0809 12:55:15.361975 139968602302272 keras_utils.py:85] BenchmarkMetric: {'global step':55800, 'time_taken': 39.636014,'examples_per_second': 5167.018027}
I0809 12:55:54.894718 139968602302272 keras_utils.py:85] BenchmarkMetric: {'global step':55900, 'time_taken': 39.532713,'examples_per_second': 5180.519747}
I0809 12:56:34.260857 139968602302272 keras_utils.py:85] BenchmarkMetric: {'global step':56000, 'time_taken': 39.366007,'examples_per_second': 5202.458038}
I0809 12:57:13.666989 139968602302272 keras_utils.py:85] BenchmarkMetric: {'global step':56100, 'time_taken': 39.406285,'examples_per_second': 5197.140564}
I0809 12:57:52.937912 139968602302272 keras_utils.py:85] BenchmarkMetric: {'global step':56200, 'time_taken': 39.270838,'examples_per_second': 5215.065722}
I0809 12:58:20.139214 139968602302272 keras_utils.py:93] BenchmarkMetric: {'epoch':89, 'time_taken': 253.945916}
625/625 - 254s - loss: 1.3511 - sparse_categorical_accuracy: 0.7797 - val_loss: 1.3852 - val_sparse_categorical_accuracy: 0.7611
```

**Comment by tfboyd** [comment]
This looks good and I am excited; a couple items for now and later:

- <now> add a benchmark and an accuracy test to the keras_imagenet_benchmarks.py.  I suggest doing both 1,8 as well as with and without XLA.  For the accuracy test you can do without XLA only and only 8 GPUs as I do not like running ResNet50 1 GPU 90 epochs.  

- <Later> if we add this to the other models (and we should) we can add this flag to /official/misc/flags/_performance.py (I think I remembered this correctly)  that way the flag is the same every where.  

I will double this up in a review comment as well.

**Comment by vinhngx** [comment]
Thanks for your prompt guidance @tfboyd. I've added a few test configs.

File: official/vision/image_classification/resnet_imagenet_main.py [hunk 1]
```diff
@@ -182,15 +185,13 @@ def run(flags_obj):
...
-          optimizer, loss_scale=flags_core.get_loss_scale(flags_obj,
-                                                          default_for_fp16=128))
-    if flags_obj.automatic_mixed_precision:
-        if dtype == 'float16':
-            raise RuntimeError("Automatic mixed precision should not be called in conjunction with "
-                               "other types of mixed precision training. Set --dtype=fp32 instead.")
-        optimizer = tf.compat.v1.train.experimental.enable_mixed_precision_graph_rewrite(optimizer)
-        
+      if flags_dict["fp16_implementation"] == "graph_rewrite":
+          optimizer = tf.compat.v1.train.experimental.enable_mixed_precision_graph_rewrite(optimizer)
```
Comment (tfboyd) [inline official/vision/image_classification/resnet_imagenet_main.py]: This should not be compat.v1, we worked really hard to get it back to experimental from purely deprecated.

File: official/transformer/v2/transformer_main.py [hunk 2]
```diff
@@ -376,6 +376,12 @@ def _create_optimizer(self):
       opt = tf.keras.mixed_precision.experimental.LossScaleOptimizer(
           opt, loss_scale=flags_core.get_loss_scale(self.flags_obj,
                                                     default_for_fp16="dynamic"))
+    if self.flags_obj.automatic_mixed_precision:
```
Comment (tfboyd) [inline official/transformer/v2/transformer_main.py]: Why is this flag here if we have the graph_rewrite flag?

Comment (vinhngx) [reply official/transformer/v2/transformer_main.py]: yes I'm moving this to using the graph_rewrite flag as well. Marking this PR as WIP till ready. Thanks.

File: official/vision/image_classification/resnet_imagenet_main.py (hunk 1 above)
Comment (vinhngx) [reply official/vision/image_classification/resnet_imagenet_main.py]: nice, thanks. Changing this to `tf.train.experimental.enable_mixed_precision_graph_rewrite`

**Comment by vinhngx** [comment]
Transformer Big 8-GPU accuracy test results via Perfzero: Model converged to  Bleu score (uncased) of 28.38 after 180k steps at batch size 3072 but appears to overfit afterwards (Bleu reduced to 27.9 after 220k steps). Throughput ~74k examples/sec. 

```
2019-08-21 07:56:50,613 INFO: Bleu score (uncased): 28.38400900363922
2019-08-21 07:56:50,613 INFO: Bleu score (cased): 27.862820029258728
2019-08-21 08:00:28,121 INFO: BenchmarkMetric: {'global step':180100, 'time_taken': 350.343444,'examples_per_second': 7014.830846}
2019-08-21 08:01:01,424 INFO: BenchmarkMetric: {'global step':180200, 'time_taken': 33.302829,'examples_per_second': 73795.531917}
2019-08-21 08:01:34,508 INFO: BenchmarkMetric: {'global step':180300, 'time_taken': 33.084607,'examples_per_second': 74282.277950}
2019-08-21 08:02:07,805 INFO: BenchmarkMetric: {'global step':180400, 'time_taken': 33.296769,'examples_per_second': 73808.963479}
2019-08-21 08:02:41,029 INFO: BenchmarkMetric: {'global step':180500, 'time_taken': 33.223781,'examples_per_second': 73971.110993}
2019-08-21 08:03:14,195 INFO: BenchmarkMetric: {'global step':180600, 'time_taken': 33.166298,'examples_per_second': 74099.315249}
2019-08-21 08:03:47,590 INFO: BenchmarkMetric: {'global step':180700, 'time_taken': 33.395007,'examples_per_second': 73591.839348}
2019-08-21 08:04:20,563 INFO: BenchmarkMetric: {'global step':180800, 'time_taken': 32.972140,'examples_per_second': 74535.653806}
2019-08-21 08:04:53,929 INFO: BenchmarkMetric: {'global step':180900, 'time_taken': 33.366377,'examples_per_second': 73654.984822}
2019-08-21 08:05:27,047 INFO: BenchmarkMetric: {'global step':181000, 'time_taken': 33.118444,'examples_per_second': 74206.384893}
```

**Comment by vinhngx** [comment]
Hi @tfboyd this is ready for another review. In summary:
- implemented mixed precision training via graph rewrite for Resnet and Transformer models via setting the flag combo `--dtype=fp16 --fp16_implementation=graph_rewrite`
- implemented perfzero test configs
- end to end accuracy tests show reasonable throughput and model accuracy.

**Comment by tfboyd** [comment]
Overfit is fine; I setup my internal test to record the best value found.

**Comment by vinhngx** [comment]
> 
> 
> Overfit is fine; I setup my internal test to record the best value found for the sanity check. It may make sense to tune the hyper parameters; but that is not something we made time to do.

ok cool. So I'll lock this down for Restnet and Transformer. 

I'm taking a look at other models later.

**Comment by tfboyd** [comment]
Looking at my calendar I hope to do some test runs today and get this approved.  I hope any small things we can just do in a followup PR.

**Comment by vinhngx** [comment]
cool, thanks @tfboyd and @saberkun

**Comment by tfboyd** [comment]
I will try to add a job for converging.  And verify the short benchmarks
were picked up automatically.  I will update this PR., So feel free to ping
me if you think of it

**Comment by tfboyd** [comment]
@vinhngx   @reedwm   

Here are the perf numbers with real data.

- AMP is: 5.2K  8xV100  753 1xV100
- Casting FP16 / Keras FP16 is:  6.8K 8xV100    895 1xV100

FP32 is:  ~2,500 8xV100 and 360 1xV100  

The numbers vary more than I would like at 8x but you get the idea.  This is not a negative as there are gains and this is just ResNet50 which is easy to do with casting. I am sharing the data only.

**Comment by vinhngx** [comment]
Thanks and noted @tfboyd. We’re working on improving this. Good thing is that any gain will be automatically passed on to users without any extra efforts :-)

**Comment by tfboyd** [comment]
I have this weird feeling maybe grappler is not switching to NHWC.  This is
a wild guess and I mentioned it to Reed.  I mean wild it is only based on
the numbers and not any actual evidence.  So even if I am right it is
because I guessed not because I know something.

**Comment by benbarsdell** [comment]
If you're using ToT then you're probably right:  https://github.com/tensorflow/tensorflow/commit/f096b13dd554ecc72986f475b4a5caa779d4ada5
undid the effect of https://github.com/tensorflow/tensorflow/pull/31087
